### PR TITLE
Improve tickets flow UI

### LINKED ISSUE #105: Use `CardBase` for cards in Tickets flow
from https://github.com/AnalogIO/coffeecard_app/pull/87#discussion_r798462541

### LINKED ISSUE #129: (Buy single drink/other products) Display ShopCard as disabled with "Coming soon" tooltip

### PR DESCRIPTION
**Pull request by marfavi**
resolves #129, resolves #105

Future:
- #213 
- #212
- #130 
- #131
- #134

### CODE REVIEW

File: lib/widgets/components/tickets/shop_card.dart
Comment (marfavi) [inline lib/widgets/components/tickets/shop_card.dart:13]: CardBase should not be a class to be inherited, use composition instead.

File: lib/widgets/components/tickets/shop_section.dart
Comment (marfavi) [inline lib/widgets/components/tickets/shop_section.dart:1]: File should be looked at

File: lib/widgets/components/tickets/swipe_overlay.dart
Comment (marfavi) [inline lib/widgets/components/tickets/swipe_overlay.dart:1]: This overlay needs to be changed.

Wrong scrim color
Use of magic number for height
Reliance on device width for slider width

File: lib/widgets/components/tickets/tickets_section.dart
Comment (marfavi) [inline lib/widgets/components/tickets/tickets_section.dart:1]: File should be looked at

File: lib/widgets/pages/tickets/buy_tickets_page.dart
Comment (marfavi) [inline lib/widgets/pages/tickets/buy_tickets_page.dart:1]: File should be looked at

File: lib/widgets/pages/tickets/tickets_page.dart
Comment (marfavi) [inline lib/widgets/pages/tickets/tickets_page.dart:1]: File should be looked at

File: lib/widgets/routers/tickets_flow.dart
Comment (marfavi) [inline lib/widgets/routers/tickets_flow.dart:1]: File should be looked at

**Comment by TTA777** [comment]
Disabled auto merge to ensure more than 1 reviewer gets the chance to look at this monster

File: lib/widgets/components/tickets/rounded_button.dart
Comment (marfavi) [reply lib/widgets/components/tickets/rounded_button.dart:1]: They should be separate (for now)

**Comment by TTA777** [comment]
@marfavi Is this ready for review? Seems you were still working on it recently?

**Comment by marfavi** [comment]
> @marfavi Is this ready for review? Seems you were still working on it recently?

adjusting some code, should be ready in 30 mn

**Comment by marfavi** [comment]
@TTA777 ready

File: lib/widgets/components/tickets/buy_ticket_bottom_modal_sheet.dart
Comment (TTA777) [inline lib/widgets/components/tickets/buy_ticket_bottom_modal_sheet.dart:32]: Personally, I would prefer that this was not a seperate class, since it is relatively simple, and only user once

File: lib/widgets/components/card.dart
Comment (TTA777) [inline lib/widgets/components/card.dart:76]: Is there any reason why a card can be disabled, and have an onTap function?

**Review by TTA777 — CHANGES_REQUESTED** [review]
May compliant is that the buy_tickets cubit seem to have regressed in the features it is offering. Few other minor comments, but I don't ahve that strong opinions when it comes to the widgets themselves

File: lib/widgets/components/tickets/buy_ticket_bottom_modal_sheet.dart
Comment (marfavi) [reply lib/widgets/components/tickets/buy_ticket_bottom_modal_sheet.dart:32]: It will be used more than one place in the near future (swiping a ticket).

Comment (TTA777) [reply lib/widgets/components/tickets/buy_ticket_bottom_modal_sheet.dart:32]: This is a private class in the buy_ticket_bottom_modal_sheet, not the BottomModalSheetHelper of almost the same exact name. Another reason to remove this one

File: lib/data/repositories/v1/product_repository.dart
Comment (marfavi) [reply lib/data/repositories/v1/product_repository.dart:22]: I think it stems from the semantics related issue #172.
This will be addressed by exposing two public methods in the cubit like this:
```dart
  /// Get purchaseable ticket (coffee card) products.
  Future<Either<ApiError, List<Product>>> getTicketProducts() async {
    return _getProducts(filter: (product) => product.amount > 1);
  }

  /// Get purchaseable single drink products.
  Future<Either<ApiError, List<Product>>> getSingleDrinkProducts() async {
    return _getProducts(filter: (product) => product.amount == 1);
  }
```

Comment (marfavi) [reply lib/data/repositories/v1/product_repository.dart:22]: I realize this adds (what could be) a redundant api call. I'll look into other solutions

Comment (TTA777) [reply lib/data/repositories/v1/product_repository.dart:22]: I believe reverting commit [2588e92](https://github.com/AnalogIO/coffeecard_app/pull/197/commits/2588e922aa86a9952d1ac8c8df94e8ae298405d0), either fully or in part, should be able to solve the issue.

File: lib/widgets/components/tickets/coffee_card.dart
Comment (marfavi) [reply lib/widgets/components/tickets/coffee_card.dart:1]: Sorry, it was meant for myself and forgot to resolve all the comments

File: lib/widgets/components/tickets/shop_card.dart
Comment (marfavi) [reply lib/widgets/components/tickets/shop_card.dart:14]: #212

File: lib/widgets/components/tickets/swipe_overlay.dart
Comment (marfavi) [reply lib/widgets/components/tickets/swipe_overlay.dart:1]: #213

File: lib/cubits/buy_tickets_page/buy_tickets_cubit.dart
Comment (marfavi) [reply lib/cubits/buy_tickets_page/buy_tickets_cubit.dart:16]: By the naming of this cubit, I thought this cubit was only meant to be used by the "Buy Tickets" flow.
I'm in favour of making this work for both flows, but the implementation of the single drink page has been postponed, so does it make sense to wait with that and focus on how it's working for the "Buy tickets" flow only?

File: lib/cubits/buy_tickets_page/buy_tickets_cubit.dart
Comment (marfavi) [reply lib/cubits/buy_tickets_page/buy_tickets_cubit.dart:9]: Removed

File: lib/data/repositories/v1/product_repository.dart
Comment (marfavi) [reply lib/data/repositories/v1/product_repository.dart]: @TTA777 What do you think of the current implementation? I added a cache for all products in the product repository.

File: lib/widgets/components/card.dart
Comment (marfavi) [reply lib/widgets/components/card.dart:76]: An oversight. Fixed

File: lib/data/repositories/v1/product_repository.dart
Comment (TTA777) [reply lib/data/repositories/v1/product_repository.dart]: Not a fan, and I don't think it would work all the time either. The product repository is registered as a factory in the service locator, so it is not guaranteed to get the same instance every time it is requested. That is, you may or may not get the cached data.

Secondly, I don't like introducing a state to the repository layer, which it essentially is. That should only be the domain of the cubits/ blocs in my eyes. Feel free to argue otherwise

File: lib/cubits/buy_tickets_page/buy_tickets_cubit.dart
Comment (TTA777) [reply lib/cubits/buy_tickets_page/buy_tickets_cubit.dart]: I agree that the naming of the cubit is weird. A better name would likely be product_cubit, since that is what it is dealing with. It does not handle the purchasing of items after all. The method name should also be renamed, since it is not getting tickets, but products. 

As for postponing the retrieval of single products, well, I suppose we can. We haven't really discussed how we want to handle the buy one drink flow recently, either. We could do it with our current endpoints, it would just take an extra request.

File: lib/widgets/components/card.dart
Comment (TTA777) [inline lib/widgets/components/card.dart:72]: Instead of having the insternal _onTap function, we could replace it with a ternary on this line like so
`onTap != null ? () => onTap!(context) : null`
What do you think?

File: lib/widgets/components/tickets/buy_ticket_bottom_modal_sheet.dart
Comment (TTA777) [inline lib/widgets/components/tickets/buy_ticket_bottom_modal_sheet.dart:77]: Weird naming, should it have been _BottomModalSheetButtonBar?

**Review by TTA777 — COMMENTED** [review]
Minor things, but I am happy with the product_cubit/ product_repository now

**Comment by marfavi** [comment]
Dart doesn't play well with rebasing again, so I'm having trouble doing it without getting bugged type errors

**Review by TTA777 — APPROVED** [review]
Lgtm, though considering the size, I think it would be good if we could get one more set of eyes on this before merging. Are you up for a look @fremartini?

**Comment by fremartini** [comment]
> Lgtm, though considering the size, I think it would be good if we could get one more set of eyes on this before merging. Are you up for a look @fremartini?

Taking a look tomorrow

**Comment by fremartini** [comment]
should the bar here not have been styled?
image

**Comment by marfavi** [comment]
> should the bar here not have been styled?
> image
> 

#213

**Review by fremartini — APPROVED** [review]

**Comment by fremartini** [comment]
@marfavi LGTM, squash and merge when you feel lile it 🙂

**Pull request merged**
